honeybadgermpc/protocols/avid.py

```diff
@@ -0,0 +1,212 @@
+                self.send(i, (sid, 'VAL', roothash_list,
+                              branch_list, stripes_list_per_party[i]))
+
+        # counters and variables for msg
+        echo_set = set()
+        ready_set = set()
+        ready_sent = False
+        from_leader = None
+        # internal storage
+        my_stripes = [None] * self.input_size
+        my_roothash_list = [None] * self.input_size
+
+        while True:  # main recv loop
+            sender, msg = await self.recv()
+            if msg[1] == 'VAL' and from_leader is None:
+                # Validation
+                (_, _, roothash_list, branch_list, stripes_for_each) = msg
+                if sender != self.leader:
+                    logger.warning(
+                        f"[{pid}] VAL message from other than leader: {sender}")
+                    continue
+
+                # merkle tree verification
+                validation_fail_flag = False
+                for i in range(len(stripes_for_each)):
+                    # verify each entry in the stripes
+                    try:
+                        assert merkle_verify(self.n, stripes_for_each[i],
+                                             roothash_list[i], branch_list[i], pid)
+                    except Exception as e:
+                        logger.info(f"[{pid}]Failed to validate VAL message: {e}")
+                        validation_fail_flag = True
+                        break
+                if validation_fail_flag:
+                    continue
+
+                # Update and store the stripes & roothashes for retrieve
+                from_leader = pid
+                my_stripes = stripes_for_each
+                my_roothash_list = roothash_list
+
+                # echo others
+                broadcast((sid, 'ECHO'))
+
+            elif msg[1] == 'ECHO':
+                # Validation
+                if sender in echo_set:
+                    logger.warning("[{pid}] Redundant ECHO")
+                    continue
+
+                # Update
+                echo_set.add(sender)
+
+                if len(echo_set) >= echo_threshold and not ready_sent:
+                    ready_sent = True
+                    broadcast((sid, 'READY'))
+
+            elif msg[1] == 'READY':
+                # Validation
+                if sender in ready_set:
+                    logger.warning("[{pid}] Redundant READY")
+                    continue
+
+                # Update
+                ready_set.add(sender)
+
+                # Amplify ready messages
+                if len(ready_set) >= ready_threshold and not ready_sent:
+                    ready_sent = True
+                    broadcast((sid, 'READY'))
+
+                if len(ready_set) >= output_threshold and len(echo_set) >= k:
+                    # update ok queue indicating ready for retrieve
+                    self.ok_queue.put_nowait(True)
+                    broadcast((sid, 'OK'))
```

Comment:
This part isn't right. This "OK" is supposed to be a local output to the caller, e.g. AVSS. There is no "OK" message sent on the network in AVID. It signifies that the AVID is complete. It's necessary to wait for this OK locally before outputting a share for AVSS.

Reply:
OK, does it mean I can just take the broadcast out? The ok queue will handle that.

Reply:
Yes you can take out the broadcast.
I think I see how `ok_queue` works.... however, I would suggest instead, making `self.ok_future` as a `Future` object. You can then call `self.ok_future.set_result(True)` instead of `self.ok_queue.put_nowait(True)`

Reply:
Yeah, that's better.